### ISSUE TITLE
new lint: pointless match

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang-nursery/rust-clippy#3337]
I wonder if it would make sense to lint against matches that match on and return the same values, like this
````rust
fn main() {
    let x = 4;
    let _ = match x {
		1 => 1,
		4 => 4,
		x => x,
	};
}
````

**Comment by matthiaskrgr** [comment]
Linting match arms where LHS == RHS might have a lot of false positives though, maybe?

**Comment by Simon-Laux** [comment]
I have another case for an pointless match statement:
```rust
fn main() {
    let mut x = 3;
    // simulate possible many lines of code using and modifying x:
    println!("x was {}", x); 
    x = 7;
    match x {
        5 => {
            println!("this can never be reached");
        }
        _ => {
            println!("this gets always executed");
        }
    }
}
```
thats a good example also for unreachable code. Although this mistake is surely more common for a machine to do than for a human, so I'm not sure if it's worth it to check that in clippy. 
Like https://github.com/jameysharp/corrode produces it sometimes something like
```rust
let mut x = 545;
    match y {
        true => {
            x = 762;
            match x {
                762 => {
                    println!("this gets always executed");
                }
                _ => {
                    println!("this can never be reached");
                }
            }
        }
        false => {
            x = 444;
            match x {
                7 => {
                    println!("this can never be reached");
                }
                _ => {
                    println!("this gets always executed");
                }
            }
        }
    }
```

**Comment by flip1995** [comment]
To check this in Clippy, we would need the information of the MIR constant propagation for variables. Sadly we don't have them (yet).

**Comment by nk9** [comment]
Do we have the necessary information now?

@rustbot label A-lint
